### ISSUE TITLE
Deduce base class function types/names from inheriting ones

### ISSUE THREAD

**Issue by jrmuizel** [astrelsky/Ghidra-Cpp-Class-Analyzer#28]
```c++
struct Bar {
        virtual int a(int x) = 0;
        virtual int b(double x, double y) = 0;
        virtual int get() = 0;
};

struct Foo : public Bar {
        int a(int ax) override {
                x = ax;
                return x;
        }
        int b(double ax, double ay) override {
                x = ax;
                y = ay;
                return y;
        }
        virtual int get() {
                return y;
        }
        int x;
        int y;
};

int main() {
        Bar* b = new Foo;
        b->a(4);
        b->b(6, 7);
        return b->get();
}
```
is currently decompiled as:
```c
void main(void)

{
  code *pcVar1;
  Foo *this;
  
  this = (Foo *)operator.new(0x10);
  Foo::Foo(this);
  pcVar1 = (code *)((this->super_Bar)._vptr)->__cxa_pure_virtual;
  (*pcVar1)(this,4,pcVar1);
  (*(code *)((this->super_Bar)._vptr)->__cxa_pure_virtual)
            (0x4018000000000000,0x401c000000000000,this);
  (*(code *)((this->super_Bar)._vptr)->__cxa_pure_virtual)(this);
  return;
}
```

It should be possible to improve the info in the Bar vtable struct by using the types and names from the Foo vtable struct.

**Comment by astrelsky** [comment]
This is doable. I will add it as an option to the class analyzers. The function names will only be used if they are not the default `FUN_` names.

**Comment by garyttierney** [comment]
Hey @astrelsky, have you done any work on this at all? I'd be happy to have a go at this and also maybe making the ClassTypeInfo window show the vftable entries of the class being looked at.

**Comment by astrelsky** [comment]
> Hey @astrelsky, have you done any work on this at all? I'd be happy to have a go at this and also maybe making the ClassTypeInfo window show the vftable entries of the class being looked at.

I haven't done so yet. You are more than - welcome to give it a shot. Collecting all the abstract classes and using `ClassTypeInfo::isAbstract` as a filter should give you a head start.

`hashCode` is implemented using the typename's `hashCode` but if you want to use a red black based collection I'd recommend using `(a, b) -> Long.compare(a.getKey(), b.getKey())` as the comparator.

Those are the only thoughts I've had yet. I started setting up nodes for the graphing but vscode kept "not responding" so I stopped working on Java stuff until it is resolved.
